// @keetanetwork/asn1-napi-rs/index.js
const { existsSync, readFileSync } = require('fs')
const { join } = require('path')

const { platform, arch } = process

let nativeBinding = null
let localFileExisted = false
let loadError = null

function isMusl() {
  // For Node 10
  if (!process.report || typeof process.report.getReport !== 'function') {
    try {
      const lddPath = require('child_process')
        .execSync('which ldd')
        .toString()
        .trim()
      return readFileSync(lddPath, 'utf8').includes('musl')
    } catch (e) {
      return true
    }
  } else {
    const { glibcVersionRuntime } = process.report.getReport().header
    return !glibcVersionRuntime
  }
}

switch (platform) {
  case 'android':
    switch (arch) {
      case 'arm64':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.android-arm64.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.android-arm64.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-android-arm64')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.android-arm-eabi.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.android-arm-eabi.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-android-arm-eabi')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Android ${arch}`)
    }
    break
  case 'win32':
    switch (arch) {
      case 'x64':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.win32-x64-msvc.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.win32-x64-msvc.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-win32-x64-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'ia32':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.win32-ia32-msvc.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.win32-ia32-msvc.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-win32-ia32-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm64':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.win32-arm64-msvc.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.win32-arm64-msvc.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-win32-arm64-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Windows: ${arch}`)
    }
    break
  case 'darwin':
    localFileExisted = existsSync(
      join(__dirname, 'asn1-napi-rs.darwin-universal.node'),
    )
    try {
      if (localFileExisted) {
        nativeBinding = require('./asn1-napi-rs.darwin-universal.node')
      } else {
        nativeBinding = require('@keetanetwork/asn1-napi-rs-darwin-universal')
      }
      break
    } catch {}
    switch (arch) {
      case 'x64':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.darwin-x64.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.darwin-x64.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-darwin-x64')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm64':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.darwin-arm64.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.darwin-arm64.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-darwin-arm64')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on macOS: ${arch}`)
    }
    break
  case 'freebsd':
    if (arch !== 'x64') {
      throw new Error(`Unsupported architecture on FreeBSD: ${arch}`)
    }
    localFileExisted = existsSync(
      join(__dirname, 'asn1-napi-rs.freebsd-x64.node'),
    )
    try {
      if (localFileExisted) {
        nativeBinding = require('./asn1-napi-rs.freebsd-x64.node')
      } else {
        nativeBinding = require('@keetanetwork/asn1-napi-rs-freebsd-x64')
      }
    } catch (e) {
      loadError = e
    }
    break
  case 'linux':
    switch (arch) {
      case 'x64':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'asn1-napi-rs.linux-x64-musl.node'),
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./asn1-napi-rs.linux-x64-musl.node')
            } else {
              nativeBinding = require('@keetanetwork/asn1-napi-rs-linux-x64-musl')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'asn1-napi-rs.linux-x64-gnu.node'),
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./asn1-napi-rs.linux-x64-gnu.node')
            } else {
              nativeBinding = require('@keetanetwork/asn1-napi-rs-linux-x64-gnu')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 'arm64':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'asn1-napi-rs.linux-arm64-musl.node'),
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./asn1-napi-rs.linux-arm64-musl.node')
            } else {
              nativeBinding = require('@keetanetwork/asn1-napi-rs-linux-arm64-musl')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'asn1-napi-rs.linux-arm64-gnu.node'),
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./asn1-napi-rs.linux-arm64-gnu.node')
            } else {
              nativeBinding = require('@keetanetwork/asn1-napi-rs-linux-arm64-gnu')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 'arm':
        localFileExisted = existsSync(
          join(__dirname, 'asn1-napi-rs.linux-arm-gnueabihf.node'),
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./asn1-napi-rs.linux-arm-gnueabihf.node')
          } else {
            nativeBinding = require('@keetanetwork/asn1-napi-rs-linux-arm-gnueabihf')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Linux: ${arch}`)
    }
    break
  default:
    throw new Error(`Unsupported OS: ${platform}, architecture: ${arch}`)
}

if (!nativeBinding) {
  if (loadError) {
    throw loadError
  }
  throw new Error(`Failed to load native binding`)
}

const {
  ASN1Decoder,
  ASN1Encoder,
  ASN1Iterator,
  ASN1BigIntToBuffer,
  BufferToBigInt,
  ASN1IntegerToBigInt,
  StringToBigInt,
  JStoASN1,
  ASN1toJS,
} = nativeBinding

module.exports.ASN1Decoder = ASN1Decoder
module.exports.ASN1Encoder = ASN1Encoder
module.exports.ASN1Iterator = ASN1Iterator
module.exports.ASN1BigIntToBuffer = ASN1BigIntToBuffer
module.exports.BufferToBigInt = BufferToBigInt
module.exports.ASN1IntegerToBigInt = ASN1IntegerToBigInt
module.exports.StringToBigInt = StringToBigInt
module.exports.JStoASN1 = JStoASN1
module.exports.ASN1toJS = ASN1toJS
